node/src/components/small_network.rs

```diff
@@ -500,6 +525,41 @@ where
         effects
     }
 
+    /// Marks connections as asymmetric (only incoming or only outgoing) and removes them if they
+    /// pass the upper limit for this. Connections that are symmetrical are reset to 0.
+    fn enforce_symmetric_connections(&mut self) {
+        if self.outgoing.is_empty() && self.incoming.is_empty() {
+            return;
+        }
+        let mut remove = Vec::new();
+        for (node_id, conn) in self.incoming.iter_mut() {
+            if !self.outgoing.contains_key(node_id) {
+                if conn.seen_asymmetrically >= MAX_ASYMMETRIC_CONNECTION_SEEN {
+                    remove.push((*node_id, conn.peer_address));
+                } else {
+                    conn.seen_asymmetrically += 1;
+                }
+            } else {
+                conn.seen_asymmetrically = 0;
+            }
+        }
+        for (node_id, conn) in self.outgoing.iter_mut() {
+            if !self.incoming.contains_key(node_id) {
+                if conn.seen_asymmetrically >= MAX_ASYMMETRIC_CONNECTION_SEEN {
+                    remove.push((*node_id, conn.peer_address));
+                } else {
+                    conn.seen_asymmetrically += 1;
+                }
+            } else {
+                conn.seen_asymmetrically = 0;
+            }
+        }
+        for (node_id, peer_address) in remove {
+            self.blocklist.insert(peer_address);
```

Comment:
I'm not sure it's worthwhile adding an incoming peer's full address here.

If a peer attempts to reconnect, chances are its address will have a different port.  We could look to block based on IP only, but that might be problematic too if we expect to continue to support multiple nodes using a single IP.

I'd be inclined to only add the outgoing addresses here, since they should represent the peer's listening address and will remain static for the given peer's session.